stan/math/rev/fun/add.hpp

```diff
@@ -0,0 +1,538 @@
+    matrix_d adjAplusB(rows_, cols_);
+    adjAplusB = Map<matrix_vi>(variRefAplusB_, rows_, cols_).adj();
+    Map<matrix_vi>(variRefB_, rows_, cols_).adj() += adjAplusB;
+  }
+};
+
+/**
+ * This is a subclass of the vari class for matrix
+ * addition A + B, where A is an N by M matrix of vars
+ * and B is an N by M matrix of doubles.
+ *
+ * The class stores the structure of each matrix,
+ * the double values of A and B, and pointers to
+ * the varis for A and B if A or B is a var. It
+ * also instatiates and stores pointers to varis
+ * for all elements of A + B.
+ *
+ * @tparam Ta type of elements in matrix A
+ * @tparam R number of rows (common to A and B)
+ * @tparam C number of columns (common to A and B)
+ */
+
+template <typename Ta, int R, int C>
+class add_mat_vari<Ta, double, R, C> : public vari {
+ public:
+  int rows_;
+  int cols_;
+  int size_;
+  double* Ad_;
+  double* Bd_;
+  vari** variRefA_;
+  vari** variRefAplusB_;
+
+  /**
+   * Constructor for add_mat_vari.
+   *
+   * All memeory is allocated in ChainableStack's
+   * stack_alloc arena.
+   *
+   * It is critical for the efficiency of this object
+   * that the constructor create new varis that aren't
+   * popped onto the var_stack_, but rather are
+   * popped onto the var_nochain_stack_. This is
+   * controlled by the second argument to
+   * vari's constructor.
+   *
+   * @param A matrix
+   * @param  B matrix
+   **/
+  add_mat_vari(const Eigen::Matrix<Ta, R, C>& A,
+               const Eigen::Matrix<double, R, C>& B)
+      : vari(0.0),
+        rows_(A.rows()),
+        cols_(B.cols()),
+        size_(A.size()),
+        Ad_(ChainableStack::instance_->memalloc_.alloc_array<double>(size_)),
+        Bd_(ChainableStack::instance_->memalloc_.alloc_array<double>(size_)),
+        variRefA_(
+            ChainableStack::instance_->memalloc_.alloc_array<vari*>(size_)),
+        variRefAplusB_(
+            ChainableStack::instance_->memalloc_.alloc_array<vari*>(size_)) {
+    using Eigen::Map;
+    Map<matrix_vi>(variRefA_, rows_, cols_) = A.vi();
+    Map<matrix_d> Ad(Ad_, rows_, cols_);
+    Map<matrix_d> Bd(Bd_, rows_, cols_);
+    Ad = A.val();
+    Bd = B;
+    Map<matrix_vi>(variRefAplusB_, rows_, cols_)
+        = (Ad + Bd).unaryExpr([](double x) { return new vari(x, false); });
+  }
+
+  virtual void chain() {
+    using Eigen::Map;
+    matrix_d adjAplusB(rows_, cols_);
+    adjAplusB = Map<matrix_vi>(variRefAplusB_, rows_, cols_).adj();
+    Map<matrix_vi>(variRefA_, rows_, cols_).adj() += adjAplusB;
+  }
+};
+
+/**
+ * Return the sum of two matrices
+ * @tparam Mat1 type of first matrix
+ * @tparam Mat2 type of second matrix
+ *
+ * @param[in] A matrix
+ * @param[in] B matrix
+ * @return Sum of two matrices
+ */
+template <typename Mat1, typename Mat2,
+          require_all_eigen_t<Mat1, Mat2>* = nullptr,
+          require_any_eigen_vt<is_var, Mat1, Mat2>* = nullptr>
+auto add(const Mat1& A, const Mat2& B) {
+  using Ta = value_type_t<Mat1>;
```

Comment:
Given that these have generalised input (i.e. can take Eigen expressions), the inputs should be passed to ```Eigen::Ref``` first, and the ```Ref``` object used throughout:

```
const Eigen::Ref<const plain_type_t<Mat1>>& A_Ref = A;
```
And then ```A_Ref``` used in the rest of the function. Otherwise, if an Eigen expression is passed as an input, it could evaluated multiple times